app/gui/src/presenter/graph/widgets.rs

```diff
@@ -0,0 +1,352 @@
+            );
+            output.widget_data <+ updates_from_cache;
+            eval input.retain_node_expressions(((node_id, expr_ids)) {
+                model.borrow_mut().retain_node_expressions(*node_id, expr_ids)
+            });
+            eval input.remove_node((node_id) model.borrow_mut().remove_node(*node_id));
+        };
+
+        let out_widget_data = output.widget_data.clone_ref();
+        let weak = Rc::downgrade(&model);
+        spawn_stream_handler(weak, manager_notifications, move |notification, model| {
+            let data = model.borrow_mut().handle_notification(notification);
+            if let Some(data) = data {
+                out_widget_data.emit(data);
+            }
+            std::future::ready(())
+        });
+
+        Self { frp, model }
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+/// Model of the Widgets controller. Manages the widget queries.
+#[derive(Debug)]
+pub struct Model {
+    manager:         Rc<Manager>,
+    graph:           ExecutedGraph,
+    widgets_of_node: HashMap<ViewNodeId, Vec<ast::Id>>,
+    widget_queries:  HashMap<ast::Id, QueryData>,
+}
+
+/// Data of the query that
+#[derive(Debug)]
+struct QueryData {
+    node_id:         ViewNodeId,
+    call_expression: ast::Id,
+    method_name:     ImString,
+    arguments:       Vec<ImString>,
+    last_update:     Option<WidgetUpdates>,
+}
+
+/// Definition of a widget request. Defines the operation expression that the widgets will be
+/// attached to.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct Request {
+    /// The node id of a node that contains the operation expression.
+    pub node_id:           ViewNodeId,
+    /// Expression of the whole method call.
+    pub call_expression:   ast::Id,
+    /// Expression of the call target. Used as a visualization target.
+    pub target_expression: ast::Id,
+    /// The suggestion id of the method that this call refers to.
+    pub call_suggestion:   SuggestionId,
+}
+
+impl Model {
+    fn handle_notification(
+        &mut self,
+        notification: Notification,
+    ) -> Option<(ViewNodeId, WidgetUpdates)> {
+        match notification {
+            Notification::ValueUpdate { target, data, .. } => {
+                let query_data = self.widget_queries.get_mut(&target)?;
+                let args: Vec<(String, serde_json::Value)> = serde_json::from_slice(&data).ok()?;
+                let updates = args
+                    .into_iter()
+                    .map(|(argument_name, meta_json)| {
+                        // Treat failed deserialization as non-widget data. The returned annotation
+                        // can be an arbitrary structure, so there will be cases when we fail to
+                        // deserialize it, as it may not contain the widget data in the first place.
+                        let deserialized = serde_json::from_value(meta_json).ok();
```

Comment:
we should log warning if we dont deserialize it. Otherwise, if engine changes protocol we might not catch errors.